### PR TITLE
feat:Update flow states names

### PR DESCRIPTION
**Pull request by dharmendrak**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/CMC-1435

### Change description ###

Rename flow states names

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/camunda/caseevents/ClaimIssueCallbackHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    private final LocalDateTime deadline = now().atTime(MIDNIGHT);
+
+    @BeforeEach
+    void setup() {
+        when(deadlinesCalculator.addMonthsToDateAtMidnight(eq(4), any(LocalDate.class)))
+            .thenReturn(deadline);
+    }
+
+    @Test
+    void shouldAddClaimNotificationDeadline() {
```
Comment (sabahirfan) [inline src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/camunda/caseevents/ClaimIssueCallbackHandlerTest.java]: Let's be consistent in naming convention as agreed, `shouldXXX_when_YYY() {}`

File: src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/camunda/caseevents/ProceedOfflineCallbackHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+})
+class ProceedOfflineCallbackHandlerTest extends BaseCallbackHandlerTest {
+
+    @Autowired
+    ProceedOfflineCallbackHandler proceedOfflineCallbackHandler;
+    @Autowired
+    private ObjectMapper objectMapper;
+
+    @Test
+    void shouldCaptureTakenOfflineDate() {
```
Comment (sabahirfan) [inline src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/camunda/caseevents/ProceedOfflineCallbackHandlerTest.java]: Let's be consistent in naming convention as agreed, `shouldXXX_when_YYY() {}`

File: src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/camunda/docmosis/GenerateClaimFormCallbackHandlerTest.java [hunk 3]
```diff
@@ -78,22 +68,19 @@
...
         when(time.now()).thenReturn(issueDate.atStartOfDay());
     }
 
     @Nested
     class AboutToSubmitCallback {
 
         @Test
-        void shouldGenerateDocumentAndSetStateAsCaseIssued_whenRespondentIsRepresented() {
-            CaseData caseData = CaseDataBuilder.builder().atStateAwaitingCaseNotification()
+        void shouldGenerateDocument() {
```
Comment (sabahirfan) [inline src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/camunda/docmosis/GenerateClaimFormCallbackHandlerTest.java]: same here please

File: src/main/java/uk/gov/hmcts/reform/unspec/model/CaseData.java [hunk 4]
```diff
@@ -126,5 +126,6 @@ public boolean hasNoOngoingBusinessProcess() {
     private final LocalDateTime applicant1ResponseDeadline;
     private final LocalDateTime applicant1ResponseDate;
     private final LocalDateTime takenOfflineDate;
+    private final LocalDateTime takenOfflineByStaffDate;
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/model/CaseData.java:130]: This is new case field `takenOfflineByStaffDate` introduced, I don't see this set any where in the code.

**Review by sabahirfan — COMMENTED** [review]
The case event `TAKE_CASE_OFFLINE` in `FlowStateAllowedEventService` is allowed only after `FULL_DEFENCE` internal state. That's not correct, I think we need to correct this as well.

File: src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/FlowPredicate.java [hunk 5]
```diff
@@ -95,20 +139,27 @@
     public static final Predicate<CaseData> claimTakenOffline = caseData ->
         caseData.getCcdState() == PROCEEDS_IN_HERITAGE_SYSTEM;
 
+    public static final Predicate<CaseData> takenOfflineBySystem = caseData ->
+        caseData.getTakenOfflineDate() != null;
+
+    public static final Predicate<CaseData> takenOfflineByStaff = caseData ->
+        caseData.getTakenOfflineByStaffDate() != null;
+
     public static final Predicate<CaseData> caseProceedsInCaseman = caseData ->
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/FlowPredicate.java]: This one is not used anywhere now except in it's unit tests. Both of the predicates `takenOfflineByStaff` & `caseProceedsInCaseman` appears to be for the same event `CASE_PROCEEDS_IN_CASEMAN`

File: src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java [hunk 6]
```diff
@@ -70,81 +79,83 @@
...
+                .transitionTo(TAKEN_OFFLINE_UNREPRESENTED_DEFENDANT).onlyIf(takenOfflineBySystem)
+            .state(PENDING_CLAIM_ISSUED_UNREGISTERED_DEFENDANT)
+                .transitionTo(TAKEN_OFFLINE_UNREGISTERED_DEFENDANT).onlyIf(takenOfflineBySystem)
+            .state(CLAIM_ISSUED)
+                .transitionTo(CLAIM_NOTIFIED).onlyIf(claimNotified)
+                .transitionTo(TAKEN_OFFLINE_BY_STAFF).onlyIf(takenOfflineByStaff)
+                .transitionTo(CLAIM_DISMISSED_PAST_CLAIM_NOTIFICATION_DEADLINE).onlyIf(pastClaimNotificationDeadline)
+            .state(CLAIM_NOTIFIED)
+                .transitionTo(CLAIM_DETAILS_NOTIFIED).onlyIf(claimDetailsNotified)
+                .transitionTo(TAKEN_OFFLINE_BY_STAFF).onlyIf(takenOfflineByStaff)
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java]: Scenario: Claim taken offline after CLAIM_NOTIFIED. state flow engine in error! 

`Ambiguous transitions permitting state [MAIN.CLAIM_ISSUED] to move to more than one next states [MAIN.CLAIM_NOTIFIED,MAIN.TAKEN_OFFLINE_BY_STAFF].` 

Please add this test and it should pass.
```
 @Test
        void shouldReturnProceedsWithOfflineJourney_whenCaseDataIsCaseProceedsInCaseman() {
            CaseData caseData = CaseDataBuilder.builder().atStateClaimNotified()
                .claimProceedsInCaseman()
                .build();

            StateFlow stateFlow = stateFlowEngine.evaluate(caseData);

            assertThat(stateFlow.getState())
                .extracting(State::getName)
                .isNotNull()
                .isEqualTo(TAKEN_OFFLINE_BY_STAFF.fullName());
            assertThat(stateFlow.getStateHistory())
                .hasSize(7)
                .extracting(State::getName)
                .containsExactly(
                    DRAFT.fullName(), CLAIM_SUBMITTED.fullName(), CLAIM_ISSUED_PAYMENT_SUCCESSFUL.fullName(),
                    PENDING_CLAIM_ISSUED.fullName(), CLAIM_ISSUED.fullName(), CLAIM_NOTIFIED.fullName(), TAKEN_OFFLINE_BY_STAFF.fullName()
                );
        }
```

File: src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java [hunk 7]
```diff
@@ -70,81 +79,83 @@
...
-                .transitionTo(CASE_PROCEEDS_IN_CASEMAN).onlyIf(caseProceedsInCaseman)
-            .state(RESPONDENT_FULL_DEFENCE)
+            .state(CLAIM_DETAILS_NOTIFIED)
+                .transitionTo(CLAIM_DETAILS_NOTIFIED_TIME_EXTENSION).onlyIf(claimDetailsNotifiedTimeExtension)
+                .transitionTo(NOTIFICATION_ACKNOWLEDGED).onlyIf(notificationAcknowledged)
+                .transitionTo(FULL_DEFENCE).onlyIf(fullDefenceAfterNotifyDetails)
+                .transitionTo(FULL_ADMISSION).onlyIf(fullAdmissionAfterNotifyDetails)
+                .transitionTo(PART_ADMISSION).onlyIf(partAdmissionAfterNotifyDetails)
+                .transitionTo(COUNTER_CLAIM).onlyIf(counterClaimAfterNotifyDetails)
+                .transitionTo(TAKEN_OFFLINE_BY_STAFF).onlyIf(takenOfflineByStaff)
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java]: let's have a unit test for that.
```
@Test
        void shouldReturnProceedsWithOfflineJourney_whenCaseProceedsInCasemanAfterCaseDetailsNotified() {
        ...
        }

File: src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java [hunk 8]
```diff
@@ -70,81 +79,83 @@
...
+                .transitionTo(CLAIM_DISMISSED_PAST_CLAIM_DISMISSED_DEADLINE).onlyIf(caseDismissed)
+            .state(CLAIM_DETAILS_NOTIFIED_TIME_EXTENSION)
+                .transitionTo(NOTIFICATION_ACKNOWLEDGED).onlyIf(notificationAcknowledgedTimeExtension)
+                .transitionTo(FULL_DEFENCE).onlyIf(fullDefence)
+                .transitionTo(FULL_ADMISSION).onlyIf(fullAdmission)
+                .transitionTo(PART_ADMISSION).onlyIf(partAdmission)
+                .transitionTo(COUNTER_CLAIM).onlyIf(counterClaim)
+                .transitionTo(TAKEN_OFFLINE_BY_STAFF).onlyIf(takenOfflineByStaff)
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java]: let's have a unit test for that.
```
@Test
        void shouldReturnProceedsWithOfflineJourney_whenCaseProceedsInCasemanAfterClaimDetailsNotifiedTimeExtension() {
        ...
        }

File: src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java [hunk 9]
```diff
@@ -70,81 +79,83 @@
...
+            .state(NOTIFICATION_ACKNOWLEDGED)
+                .transitionTo(NOTIFICATION_ACKNOWLEDGED_TIME_EXTENSION)
+                    .onlyIf(notificationAcknowledgedTimeExtension)
+                .transitionTo(FULL_DEFENCE).onlyIf(fullDefenceAfterAcknowledge)
+                .transitionTo(FULL_ADMISSION).onlyIf(fullAdmissionAfterAcknowledge)
+                .transitionTo(PART_ADMISSION).onlyIf(partAdmissionAfterAcknowledge)
+                .transitionTo(COUNTER_CLAIM).onlyIf(counterClaimAfterAcknowledge)
+                .transitionTo(TAKEN_OFFLINE_BY_STAFF).onlyIf(takenOfflineByStaff)
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java]: let's have a unit test for that.
```
@Test
        void shouldReturnProceedsWithOfflineJourney_whenCaseProceedsInCasemanAfterNotificationAcknowledged() {
        ...
        }

File: src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java [hunk 10]
```diff
@@ -70,81 +79,83 @@
...
+                .transitionTo(CLAIM_DISMISSED_PAST_CLAIM_DISMISSED_DEADLINE).onlyIf(caseDismissedAfterClaimAcknowledged)
+            .state(NOTIFICATION_ACKNOWLEDGED_TIME_EXTENSION)
+                .transitionTo(FULL_DEFENCE).onlyIf(fullDefence)
+                .transitionTo(FULL_ADMISSION).onlyIf(fullAdmission)
+                .transitionTo(PART_ADMISSION).onlyIf(partAdmission)
+                .transitionTo(COUNTER_CLAIM).onlyIf(counterClaim)
+                .transitionTo(TAKEN_OFFLINE_BY_STAFF).onlyIf(takenOfflineByStaff)
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java]: let's have a unit test for that.
```
@Test
        void shouldReturnProceedsWithOfflineJourney_whenCaseProceedsInCasemanAfterNotificationAcknowledgedTimeExtension() {
        ...
        }

File: src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java [hunk 11]
```diff
@@ -70,81 +79,83 @@
...
+                .transitionTo(CLAIM_DISMISSED_PAST_CLAIM_DISMISSED_DEADLINE).onlyIf(caseDismissedAfterClaimAcknowledged)
+            .state(FULL_DEFENCE)
                 .transitionTo(FULL_DEFENCE_PROCEED).onlyIf(fullDefenceProceed)
                 .transitionTo(FULL_DEFENCE_NOT_PROCEED).onlyIf(fullDefenceNotProceed)
-                .transitionTo(CLAIM_DISCONTINUED).onlyIf(claimDiscontinued)
-                .transitionTo(CLAIM_WITHDRAWN).onlyIf(claimWithdrawn)
-                .transitionTo(CASE_PROCEEDS_IN_CASEMAN).onlyIf(caseProceedsInCaseman)
+                .transitionTo(TAKEN_OFFLINE_BY_STAFF).onlyIf(takenOfflineByStaff)
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java:155]: let's have a unit test for that.
```
@Test
        void shouldReturnProceedsWithOfflineJourney_whenCaseProceedsInCasemanAfterFullDefence() {
        ...
        }

File: src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java [hunk 12]
```diff
@@ -70,81 +79,83 @@
...
-                .transitionTo(CLAIM_WITHDRAWN).onlyIf(claimWithdrawn)
-                .transitionTo(CASE_PROCEEDS_IN_CASEMAN).onlyIf(caseProceedsInCaseman)
-                .transitionTo(PROCEEDS_OFFLINE_ADMIT_OR_COUNTER_CLAIM).onlyIf(claimTakenOffline)
-            .state(RESPONDENT_COUNTER_CLAIM)
-                .transitionTo(CLAIM_DISCONTINUED).onlyIf(claimDiscontinued)
-                .transitionTo(CLAIM_WITHDRAWN).onlyIf(claimWithdrawn)
-                .transitionTo(CASE_PROCEEDS_IN_CASEMAN).onlyIf(caseProceedsInCaseman)
-                .transitionTo(PROCEEDS_OFFLINE_ADMIT_OR_COUNTER_CLAIM).onlyIf(claimTakenOffline)
+            .state(FULL_ADMISSION)
+                .transitionTo(TAKEN_OFFLINE_BY_STAFF).onlyIf(takenOfflineByStaff)
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java:158]: I don't think this transition is required, please double check with @ckalista

File: src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java [hunk 13]
```diff
@@ -70,81 +79,83 @@
...
+            .state(PART_ADMISSION)
+                .transitionTo(TAKEN_OFFLINE_BY_STAFF).onlyIf(takenOfflineByStaff)
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java:160]: I don't think this transition is required, please double check with @ckalista

File: src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java [hunk 14]
```diff
@@ -70,81 +79,83 @@
...
+            .state(COUNTER_CLAIM)
+                .transitionTo(TAKEN_OFFLINE_BY_STAFF).onlyIf(takenOfflineByStaff)
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java:162]: I don't think this transition is required, please double check with @ckalista

File: src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/FlowPredicate.java [hunk 15]
```diff
@@ -80,13 +131,6 @@
...
             && caseData.getTakenOfflineDate() == null;
 
-    public static final Predicate<CaseData> claimWithdrawn = caseData ->
-        caseData.getWithdrawClaim() != null
-            && caseData.getCcdState() == CASE_DISMISSED;
-
-    public static final Predicate<CaseData> respondentAgreedExtension = caseData ->
-        caseData.getRespondentSolicitor1AgreedDeadlineExtension() != null;
-
     public static final Predicate<CaseData> claimDiscontinued = caseData ->
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/FlowPredicate.java]: This is unused please remove this.

File: src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/FlowPredicate.java [hunk 16]
```diff
@@ -95,20 +139,61 @@
     public static final Predicate<CaseData> claimTakenOffline = caseData ->
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/FlowPredicate.java]: do we need this? can't see this being used anywhere except unit tests

File: src/test/java/uk/gov/hmcts/reform/unspec/sampledata/CaseDataBuilder.java [hunk 17]
```diff
@@ -645,55 +645,116 @@ public CaseDataBuilder atStatePaymentFailed() {
...
         return this;
     }
 
-    public CaseDataBuilder atStateExtensionRequested() {
-        atStateClaimAcknowledge();
-        respondentSolicitor1AgreedDeadlineExtension = LocalDate.now();
+    public CaseDataBuilder atStateClaimDetailsNotifiedTimeExtension() {
+        atStateClaimDetailsNotified();
+        respondent1ResponseDeadline = RESPONSE_DEADLINE;
         respondent1TimeExtensionDate = LocalDateTime.now();
```
Comment (sabahirfan) [inline src/test/java/uk/gov/hmcts/reform/unspec/sampledata/CaseDataBuilder.java:722]: I think we should set `respondentSolicitor1AgreedDeadlineExtension = LocalDate.now();` in this builder as well. what do you think?

File: src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java (hunk 14 above)
Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java:162]: yes, it's not required as the flow go to offline on it's own

File: src/test/java/uk/gov/hmcts/reform/unspec/sampledata/CaseDataBuilder.java (hunk 17 above)
Comment (dharmendrak) [reply src/test/java/uk/gov/hmcts/reform/unspec/sampledata/CaseDataBuilder.java:722]: RESPONSE_DEADLINE is set as now() + 14 days, which sound bit better as issue date is set as now()

Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/reform/unspec/sampledata/CaseDataBuilder.java:722]: I meant we need to set some value for respondentSolicitor1AgreedDeadlineExtension whether it’s now() or something else

File: src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java [hunk 18]
```diff
@@ -70,81 +85,82 @@
...
+                    .onlyIf(takenOfflineByStaffAfterNotificationAcknowledged)
+                .transitionTo(CLAIM_DISMISSED_PAST_CLAIM_DISMISSED_DEADLINE).onlyIf(caseDismissedAfterClaimAcknowledged)
+            .state(NOTIFICATION_ACKNOWLEDGED_TIME_EXTENSION)
+                .transitionTo(FULL_DEFENCE).onlyIf(fullDefence)
+                .transitionTo(FULL_ADMISSION).onlyIf(fullAdmission)
+                .transitionTo(PART_ADMISSION).onlyIf(partAdmission)
+                .transitionTo(COUNTER_CLAIM).onlyIf(counterClaim)
+                .transitionTo(TAKEN_OFFLINE_BY_STAFF)
+                    .onlyIf(takenOfflineByStaffAfterNotificationAcknowledgedTimeExtension)
+                .transitionTo(CLAIM_DISMISSED_PAST_CLAIM_DISMISSED_DEADLINE).onlyIf(caseDismissedAfterClaimAcknowledged)
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/flowstate/StateFlowEngine.java]: This transition will be at error state. as using the same predicate `caseDismissedAfterClaimAcknowledged` used for previous state. Below is the test case to verify that.

```@Test
        void shouldReturnClaimDismissedPastDeadline_whenCaseDataAtStateNotificationAcknowledgedTimeExtension() {
            CaseData caseData = CaseDataBuilder.builder().atStateNotificationAcknowledgedTimeExtension()
                .claimDismissedDate(LocalDateTime.now())
                .build();

            StateFlow stateFlow = stateFlowEngine.evaluate(caseData);

            assertThat(stateFlow.getState())
                .extracting(State::getName)
                .isNotNull()
                .isEqualTo(CLAIM_DISMISSED_PAST_CLAIM_DISMISSED_DEADLINE.fullName());
            assertThat(stateFlow.getStateHistory())
                .hasSize(9)
                .extracting(State::getName)
                .containsExactly(
                    DRAFT.fullName(),
                    CLAIM_SUBMITTED.fullName(), CLAIM_ISSUED_PAYMENT_SUCCESSFUL.fullName(),
                    PENDING_CLAIM_ISSUED.fullName(), CLAIM_ISSUED.fullName(), CLAIM_NOTIFIED.fullName(),
                    CLAIM_DETAILS_NOTIFIED.fullName(), NOTIFICATION_ACKNOWLEDGED.fullName(),
                    NOTIFICATION_ACKNOWLEDGED_TIME_EXTENSION.fullName(),
                    CLAIM_DISMISSED_PAST_CLAIM_DISMISSED_DEADLINE.fullName()
                );
        }
        ````

**Review by sabahirfan — DISMISSED** [review]

**Review by sabahirfan — APPROVED** [review]

**Review by douglasrice — APPROVED** [review]

**Pull request merged**